_posts/2020-10-27-a-closer-look-at-the-interpreter.md

```diff
@@ -0,0 +1,202 @@
+    jmp    [rbx]
+</pre>
+
+The bold section is the meat of the instruction and the rest is argument
+unpacking and instruction dispatch. While this is not much of a problem for
+large instructions, its effect on short ones like this is very large, and when
+looking through a profiler (e.g. [`perf`]) it's not unusual for the final `jmp`
+to dominate the rest.
+
+To lessen this effect, the loader combines commonly used instruction sequences
+into a single instruction. For example, two independent moves may fuse into
+[`move2_par`], and `is_nonempty_list` followed by `get_list` might be fused to
+[`is_nonempty_list_get_list`].
+
+This reduces the cost of short instructions, but only works to the extent we're
+able to identify common patterns and comes at a significant maintenance cost
+as each combination must be implemented manually. Even so, the effect tends to
+be moderate and the dispatch overhead remains significant.
+
+Another, albeit lesser, downside with the interpreter is that modern processors
+are very optimized for patterns commonly found in "ordinary" native code. For
+example, nearly all of them have a special branch predictor just for calls and
+returns. Assuming that every call has a corresponding return lets it predict
+returns _perfectly_ unless an exception is thrown, but since the interpreter
+does not use native calls and returns, it cannot make use of this optimization.
+
+Unfortunately there's not a whole lot that can be done about this, and after
+over two decades of refinement it's becoming increasingly difficult to optimize
+it in meaningful ways.
+
+Because of that our quest to improve performance has instead focused on two
+areas; improving the compiler, and implementing a JIT. We've made great strides
+with both as of late, and are very proud to have finally merged the latter.
+Let's have a peek at how we generate this instruction in our JIT:
+
+```
+void BeamModuleAssembler::emit_is_nonempty_list(const ArgVal &Fail,
+                                                const ArgVal &Src) {
+    /* Figure out which register to look at. This doesn't
+     * result in an instruction, it merely changes the
+     * addressing used in the `test` instruction. */
+    x86:Mem list_ptr = getArgRef(Src);
+
+    a.test(list_ptr, imm(_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST));
+    a.jne(labels[Fail.getValue()]);
+}
+```
+
+Notice how it's almost identical to the assembly listing from interpreter,
+except there's no longer any code for instruction dispatch or argument
+handling. Stay tuned for our next post, where we'll take a look at the JIT
+itself.
+
+[directly threaded code]: https://en.wikipedia.org/wiki/Threaded_code
+[`ops.tab`]: https://github.com/erlang/otp/blob/OTP-23.1/erts/emulator/beam/ops.tab
+[`beam_makeops`]: https://github.com/erlang/otp/blob/OTP-23.1/erts/emulator/utils/beam_makeops
+[documentation for the `beam_makeops` script]: https://github.com/erlang/otp/blob/OTP-23.1/erts/emulator/internal_doc/beam_makeops.md
```

Comment:
I suggest that you link to the built documentation: http://erlang.org/doc/apps/erts/beam_makeops.html